### ISSUE TITLE
Add Gitea support

### ISSUE THREAD

**Issue by ijsnow** [sourcegraph/sourcegraph#343]
#### Feature request description

[Add support for Gitea.](https://github.com/go-gitea/gitea/issues/5010) To do this, we need to implement [repo-updater](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@1d7722a5295141024f3d2b0e32be790a04827112/-/tree/cmd/repo-updater/repos) support for it.

#### Is your feature request related to a problem? If so, please describe.

Gitea users would like to use Sourcegraph!

#### Describe alternatives you've considered.

Using [repos.list config option](https://about.sourcegraph.com/docs/config/site/#reposlist-array) but maintaining this list with many repos is cumbersome.

#### Additional context

This will begin adding support to Gitea via the browser extension https://github.com/go-gitea/gitea/issues/5010. First we need to be able to clone repos from Gitea.

cc @keegancsmith

**Comment by keegancsmith** [comment]
I guess I would say that there is a lot of duplication between how we interact with code hosts in repo-updater that needs to be consolidated. So right now the best approach is to copy-paste an existing integration in repo-updater, then make it work for gitea. At some point this will be simplified, but the copy paste approach is relatively quick to do and hasn't really bitten us (yet).

Are gitea installations generally quite small? Is there a public gitea host with lots of repos? A lot of the complications in our code is avoiding hitting rate limits, which may not be an issue for gitea.

**Comment by tboerger** [comment]
Maybe drone/go-scm could also help for the abstraction of the different providers?

**Comment by keegancsmith** [comment]
@tboerger thanks for the suggestion, that actually looks fantastic. Will take a deeper look next month when I plan to tackle some of the debt that has built up in our repo syncing infra.

**Comment by issmirnov** [comment]
Hi team,

Gitea has surpassed 13,000 stars on github, so user base is likely close to 50k people. Any chance this could get added to a future milestone? Many of us would really appreciate the integration :smile: 

Thanks!

**Comment by sqs** [comment]
We haven’t heard any requests from teams above 30 yet. That’s usually the best use of our own time so we have the biggest impact on users. Anyone here interested in submitting a PR for this, or on large teams wanting this functionality?

**Comment by STaRDoGG** [comment]
Hey all, new to SG and just tried adding my locally hosted (docker) instance of Gitea to SG and can't, so searched for info on it and landed here. Any updates on this?

**Comment by systemmonkey42** [comment]
I landed here when searching for ways to support Gitea as a 'code host' for SourceGraph.  I have around 350 repositories to import, and that number is growing so a manual process is not an option.  Its, been a few years, now.  Can it really be that hard to implement?

I don't know GO, so implementing the new api was also not an option for me, instead I implemented an api proxy which I point SourceGraph to using the Github api.  It makes a gitea api call and returns the information in a way that satisfies the github importer.

Worked suprisingly well.   Does anyone think this will be a problem down the track?  Apart from verifying the user and listing repos, it there much else I would need?  I certainly cant see the github module touching issues/prs or any other github specific feature.

Thanks

**Comment by keegancsmith** [comment]
@systemmonkey42 very cool! (cc @unknwon @sourcegraph/core-application).

> Does anyone think this will be a problem down the track?

If we change our interactions with GitHub, the proxy will have to play catchup. But it's been pretty stable for a long time, so I don't see that being an issue soon.

There is a simple API we support. See the src-cli's `serve-git` command to see it. This may drastically simplify your proxy needs.

**Comment by unknwon** [comment]
Very much like what I've tried to propose in https://github.com/sourcegraph/sourcegraph/issues/5115, but sticking with GitHub APIs probably is the better approach in practice.

For the purpose of syncing repositories into Sourcegraph, I do not think you need anything else. But if you also want to enforce permissions, then additional APIs are needed, this file is the source of truth for all permissions-related APIs we use for GitHub: https://github.com/sourcegraph/sourcegraph/blob/main/internal/authz/github/github.go

We should probably consider documenting what APIs we use for each code host in terms of each function module?

**Comment by systemmonkey42** [comment]
> There is a simple API we support. See the src-cli's `serve-git` command to see it. This may drastically simplify your proxy needs.

I agree, but I know the github API, and I suspect it is less likely to make breaking changes in the forseeable future. 
TBH though, the difference between what `gitea/api/v1/user/repos` returns, and what sourcegraph expects from github is so small, the entire "conversion" is two lines (add the missing "node_id" field, and ensure "disabled" is false...  Learning enough *_go_* to read the github importer source was the hardest part :-)

The beauty of this seems to be that the URL's returned from gitea's `/usr/repos` call are enough to allow sourcegraph to clone and update repositories without ever needing to call the github api ever again.  Only the "code host" update process touches it.

 > Very much like what I've tried to propose in #5115, but sticking with GitHub APIs probably is the better approach in practice.

I suspect github has become the defacto api for repository management, with gitlab/bitbucket/gitea/etc having only minor differences.  So yes. #5115 but using github's api as the "general protocol" works, for me at least.

Hence any repository host that supports a github compatible `/user/repos` api call will work.

> For the purpose of syncing repositories into Sourcegraph, I do not think you need anything else. But if you also want to enforce permissions, then additional APIs are needed.

Yes.  Mapping github permissions to user/repositories seems to require github oauth the be enabled...  Since github isn't actually involved in any of this, I use the token provided in the "code host" config for accessing the gitea repos.  Permissions in gitea are
enforced, but not in sourcegraph which is a problem I plan to tackle at some point.

**Comment by unknwon** [comment]
> > For the purpose of syncing repositories into Sourcegraph, I do not think you need anything else. But if you also want to enforce permissions, then additional APIs are needed.
> 
> Yes. Mapping github permissions to user/repositories seems to require github oauth the be enabled... Since github isn't actually involved in any of this, I use the token provided in the "code host" config for accessing the gitea repos. Permissions in gitea are
> enforced, but not in sourcegraph which is a problem I plan to tackle at some point.

You may also push permissions to Sourcegraph: https://docs.sourcegraph.com/admin/repo/permissions#explicit-permissions-api

**Comment by techknowlogick** [comment]
FWIW, Gitea 1.15.0-dev supports the OIDC standard, and so I've been able to connect Gitea & Sourcegraph to allow users to login, and with @systemmonkey42's `gitea-api-proxy` repo above, I've been able to have a (somewhat cumbersome as an external service is still required) Sourcegraph x Gitea experience.

**Comment by techknowlogick** [comment]
Looking at the sourcegraph homepage, Gitea integration is mentioned twice. Would Sourcegraph be amenable to sponsoring a Gitea maintainer to submit a PR to Sourcegraph to support Gitea integration?

(I am one of the project leads of Gitea and am able to facilitate coordination with Gitea maintainers if the answer to above is yes)

**Comment by keegancsmith** [comment]
Also cc @tsenart who recently added pagure support, so has context on adding codehost support. Lots of codehosts on the internet use gitea and is quite a popular self hosted option, so this may help us towards our many codehosts efforts. cc @jjeffwarner

**Comment by bluecmd** [comment]
Fun fact: If number of :+1: is anything to go by, this is the highest rated issue right now :-).

**Comment by keegancsmith** [comment]
Got nerd sniped to do this with your comment @bluecmd https://github.com/sourcegraph/sourcegraph/pull/45007

**Comment by sgabenov** [comment]
Any updates on this issue? Any planned work to add support of gitea hosted repos?
